api/auth_middleware_test.go

```diff
@@ -1,63 +1,93 @@
 	tokensRepo := helpers.GetRepo(t).Tokens
-	tokens, _ := tokensRepo.FindByEmail(rootSA.Email)
+	tokens, _ := tokensRepo.FindByEmail(oauthSA.Email)
 	token := tokens[0]
-	app := helpers.GetApp(t)
-	req, err := http.NewRequest(http.MethodGet, "/service_accounts", nil)
 
-	if err != nil {
-		t.Errorf("Could not create HTTP request for /service_accounts")
+	testCases := []struct {
+		name                string
+		serviceAccount      *models.ServiceAccount
+		requestHeaders      map[string]string
+		wantResponseHeaders map[string]string
+		wantResponseCode    int
+	}{
+		{
+			name:                "KeyPairAuthorization",
+			serviceAccount:      keyPairSA,
+			requestHeaders:      map[string]string{"authorization": fmt.Sprintf("keypair %s:%s", keyPairSA.KeyID, keyPairSA.KeySecret)},
+			wantResponseHeaders: map[string]string{"x-service-account-name": "keyPairUser"},
+			wantResponseCode:    http.StatusOK,
+		},
+		{
+			name:                "OAuthAuthorization",
+			serviceAccount:      oauthSA,
+			requestHeaders:      map[string]string{"authorization": fmt.Sprintf("bearer %s", token.AccessToken)},
+			wantResponseHeaders: map[string]string{"x-email": oauthSA.Email},
+			wantResponseCode:    http.StatusOK,
+		},
+		{
+			name:             "WrongOAuthTokenAuthorization",
+			serviceAccount:   oauthSA,
+			requestHeaders:   map[string]string{"authorization": fmt.Sprintf("bearer %s", "wrong token")},
+			wantResponseCode: http.StatusUnauthorized,
+		},
+		{
+			name:             "WrongKeyPairAuthorization",
+			serviceAccount:   keyPairSA,
+			requestHeaders:   map[string]string{"authorization": fmt.Sprintf("keypair %s:%s", "wrong key_id", "wrong_key_secret")},
+			wantResponseCode: http.StatusUnauthorized,
+		},
+		{
+			name:             "UndefinedAuthorization",
+			serviceAccount:   nil,
+			wantResponseCode: http.StatusUnauthorized,
+		},
+		{
+			name:             "UnsupportedAuthorization",
+			serviceAccount:   oauthSA,
+			requestHeaders:   map[string]string{"authorization": fmt.Sprintf("basic %s", base64.StdEncoding.EncodeToString([]byte("user:password")))},
+			wantResponseCode: http.StatusUnauthorized,
+		},
 	}
 
-	req.Header.Set("Authorization", fmt.Sprintf("bearer %s", token))
-	response := helpers.DoRequest(t, req, app.GetRouter())
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			app := helpers.GetApp(t)
+			req, err := http.NewRequest(http.MethodGet, "/service_accounts", nil)
+			if err != nil {
+				t.Fatalf("Could not create HTTP request")
+			}
 
-	if response.Code != http.StatusOK {
-		t.Errorf("Expected status %d. Got %d", http.StatusOK, response.Code)
-	}
+			for header, content := range testCase.requestHeaders {
+				req.Header.Set(header, content)
+			}
+
+			response := helpers.DoRequest(t, req, app.GetRouter())
+
+			if response.Code != testCase.wantResponseCode {
+				t.Errorf("Status = %d, want %d", response.Code, testCase.wantResponseCode)
```

Comment:
nit: as mentioned above, just use %v here. It'll make the format more resilient to refactorings/type changes.